TCP refactoring

pfl opened the pull request:
This PR will move TCP functionality from net_context.c to tcp.c, while leaving TCP function calls in net_context.c that resolve to empty inline functions when TCP is not enabled. As small pieces as possible are worked on one by one, except for the last patch that moves a majority of the TCP functionality to the new file. No functionality has been changed, except for patches 4/14 and 13/14. The issue described this desired change is #5482.

Patch 4/14 reorders the function calls, so that address checks happen before offloading, and any bind calls to the stack happen after offloading.
Patch 13/14 contains a fix where calling connect will set both local and remote addresses also for a UDP.

pfl committed:
net: tcp: Define empty static functions when TCP is disabled

With CONFIG_NET_TCP is not set, provide empty static inline
prototypes for all TCP functions available to other parts of
the IP stack.

Signed-off-by: Patrik Flykt <patrik.flykt@intel.com>

pfl committed:
net: context: Update net_context_sendto()

Only #ifdefs are removed in net_context_sendto().

Signed-off-by: Patrik Flykt <patrik.flykt@intel.com>

pfl committed:
net: context: Refactor sendto()

Refactor sendto() code so that destination address and its validity
is checked first, followed by offloading verification. Move context
and shutdown checks into TCP queueing function.

Signed-off-by: Patrik Flykt <patrik.flykt@intel.com>

pfl committed:
net: context: Factor out TCP header length checking

TCP header length checking is needed by net_context.c.

Signed-off-by: Patrik Flykt <patrik.flykt@intel.com>

pfl committed:
net: context: Remove unnecessary ifdefs

Ifdefs around TCP code can be removed since the TCP code will compile
to empty functions when not enabled.

Signed-off-by: Patrik Flykt <patrik.flykt@intel.com>

pfl committed:
net: context: Factor out code for TCP receive

Create a function for TCP receive.

Signed-off-by: Patrik Flykt <patrik.flykt@intel.com>

pfl committed:
net: context: Refactor net_tcp_send_data()

Refactor net_tcp_send_data() by adding callback information in the
function.

Signed-off-by: Patrik Flykt <patrik.flykt@intel.com>

pfl committed:
net: context: Remove unnecessary callback clearing

A new context cannot have callbacks set, as it was freshly created
in the beginning of the function. Thus the extra callback clearing
can be removed. This is a left-over from commit ce41d5f432.

Signed-off-by: Patrik Flykt <patrik.flykt@intel.com>

pfl committed:
net: context: Refactor queueing of a TCP FIN when closing socket

Queue a TCP FIN packet when needed if the socket was connected or
listening and where FIN wasn't already received.

Signed-off-by: Patrik Flykt <patrik.flykt@intel.com>

pfl committed:
net: context: Factor out TCP context listen state

Factor out TCP context listening state change.

Signed-off-by: Patrik Flykt <patrik.flykt@intel.com>

pfl committed:
net: context: Factor out TCP receive window handling

Receive window is needed only for TCP.

Signed-off-by: Patrik Flykt <patrik.flykt@intel.com>

pfl committed:
net: context: Fix connect for UDP contexts

Similar to the socket connect() call, calling connect for a UDP
context needs to set both local and remote addresses and port
number. Fix this not to be exclusive for TCP.

Similarly, the remote destination can be a multicast address when
UDP is used.

Signed-off-by: Patrik Flykt <patrik.flykt@intel.com>

lpereira commented on subsys/net/ip/tcp.c:
```diff
@@ -1344,3 +1412,1160 @@ int net_tcp_parse_opts(struct net_pkt *pkt, int opt_totlen,
 	return -EINVAL;
 }
+
+int tcp_hdr_len(struct net_pkt *pkt)
```
This function used to be static; if it's not any longer, please call it `net_tcp_hdr_len()`.

pfl replied:
This function became public, as it was moved to tcp.c. With the rest of the stack moving to tcp.c, this function can again be made static.

pfl replied:
Sorry, no. tcp_hdr_len() is defined in tcp.c, but the usage is actually in net_context.c.

lpereira commented on subsys/net/ip/tcp.c:
```diff
@@ -1344,3 +1412,1160 @@ int net_tcp_parse_opts(struct net_pkt *pkt, int opt_totlen,
+int net_tcp_listen(struct net_context *context)
+{
+	if (net_context_get_ip_proto(context) == IPPROTO_TCP) {
+		net_tcp_change_state(context->tcp, NET_TCP_LISTEN);
```
Is `net_tcp_change_state()` exported? If so, can it be used only within `tcp.c`, and marked as static? Same idea for `net_tcp_get_state()`.

lpereira reviewed (commented):
I wanted to do this for a long time, so thank you for working on this.  It's a big patch and it's been a while since I've worked on the network stack, so I only have a few comments.

pfl committed:
tests: socket: Enable loopback and net tests

Enable the loopback driver and set up network testing so that it
will not require a SLIP driver. Move values from the extra test
config so that they are always enabled.

Signed-off-by: Patrik Flykt <patrik.flykt@intel.com>

pfl committed:
net: tcp: Move TCP stack to tcp.c

Move core TCP functionality from net_context.c to tcp.c. Create empty
functions that the compiler can remove if TCP is not configured. As a
result remove TCP ifdefs from net_context.

Signed-off-by: Patrik Flykt <patrik.flykt@intel.com>

tbursztyka commented on subsys/net/ip/net_context.c:
```diff
@@ -399,65 +399,74 @@ int net_context_get(sa_family_t family,
 
 #if defined(CONFIG_NET_CONTEXT_CHECK)
 
+	switch(family) {
```
Why changing that to a switch?

tbursztyka commented on subsys/net/ip/net_context.c:
```diff
@@ -399,65 +399,74 @@ int net_context_get(sa_family_t family,
+		return -EAFNOSUPPORT;
+	}
 
+	switch(type) {
```
same as above, why moving towards a switch here?

tbursztyka commented on subsys/net/ip/net_context.c:
```diff
@@ -399,65 +399,74 @@ int net_context_get(sa_family_t family,
+		break;
+
+	default:
+		NET_ASSERT_INFO(family == AF_INET || family == AF_INET6,
```
why retesting this as it is down where family cannot be AF_INET or AF_INET6 ?
Just NET_ERR/NET_WARN here no?

tbursztyka commented on subsys/net/ip/net_context.c:
```diff
@@ -399,65 +399,74 @@ int net_context_get(sa_family_t family,
+	case SOCK_DGRAM:
 #if !defined(CONFIG_NET_UDP)
-	if (type == SOCK_DGRAM) {
 		NET_ASSERT_INFO(type != SOCK_DGRAM,
```
ok so legacy code is using an ASSERT where NET_ERR/NET_WARN could have been sufficient.

tbursztyka commented on subsys/net/ip/net_context.c:
```diff
@@ -399,65 +399,74 @@ int net_context_get(sa_family_t family,
+	case SOCK_STREAM:
 #if !defined(CONFIG_NET_TCP)
-	if (type == SOCK_STREAM) {
 		NET_ASSERT_INFO(type != SOCK_STREAM,
```
same here

tbursztyka commented on subsys/net/ip/net_context.c:
```diff
@@ -399,65 +399,74 @@ int net_context_get(sa_family_t family,
-
-	if (type != SOCK_DGRAM && type != SOCK_STREAM) {
+	default:
 		NET_ASSERT_INFO(type == SOCK_DGRAM || type == SOCK_STREAM,
```
same

tbursztyka commented on subsys/net/ip/net_context.c:
```diff
@@ -399,65 +399,74 @@ int net_context_get(sa_family_t family,
 	}
 
-	if (ip_proto != IPPROTO_UDP && ip_proto != IPPROTO_TCP) {
+	switch (ip_proto) {
```
again, why moving towards switch?

tbursztyka commented on subsys/net/ip/net_context.c:
```diff
@@ -2217,17 +2217,20 @@ static void set_appdata_values(struct net_pkt *pkt, enum net_ip_protocol proto)
 	struct net_buf *frag;
 	u16_t offset;
 
+	switch(proto) {
```
again, why moving to a switch here?

tbursztyka reviewed (commented):
Original code was not using switch on purpose: so it would really get rid of an unsupported family test depending on Kconfig option. This was saving some rom. It's not much locally, but once, twice, ... n times: it saves bits. (note that a  switch on 2 choices isn't even faster than if/else, surely it's nicer to read but that's it).

So either you keep legacy code as is, or you still use #ifdef/#endif relevantly to nuke the unsupported cases from the switch (compiler will then optimize properly).

pfl commented:
Two bugs discovered and fixed, I'll update this and address the comments.

jukkar reviewed (approved):
LGTM

pfalcon commented on tests/net/socket/tcp/testcase.yaml:
```diff
@@ -1,7 +1,4 @@
 tests:
   test:
-    extra_configs:
```
The idea here was to make sure that the test works with both "normal" setup (the same as the samples are run with) and with CONFIG_NET_TEST=y. When something is made to work with only non-default setup, there's always a suspicion that some black voodoo is needed to make it work *at all*.
